storage/cloudspanner/tree_storage.go

```diff
@@ -0,0 +1,523 @@
+		glog.Warningf("no head found for treeID %v", treeID)
+		return nil, storage.ErrTreeNeedsInit
+	}
+	return th, nil
+}
+
+type newCacheFn func(*trillian.Tree) (cache.SubtreeCache, error)
+
+func (t *treeStorage) getTreeAndConfig(ctx context.Context, treeID int64, opts trees.GetOpts) (*trillian.Tree, proto.Message, error) {
+	tree, err := trees.GetTree(ctx, t.admin, treeID, opts)
+	if err != nil {
+		return nil, nil, err
+	}
+	config, err := unmarshalSettings(tree)
+	if err != nil {
+		return nil, nil, err
+	}
+	return tree, config, nil
+}
+
+// begin returns a newly started tree transaction for the specified tree.
+func (t *treeStorage) begin(ctx context.Context, treeID int64, opts trees.GetOpts, newCache newCacheFn, stx spanRead) (*treeTX, error) {
+	tree, config, err := t.getTreeAndConfig(ctx, treeID, opts)
+	if err != nil {
+		return nil, err
+	}
+	cache, err := newCache(tree)
+	if err != nil {
+		return nil, err
+	}
+	treeTX := &treeTX{
+		treeID: treeID,
+		ts:     t,
+		stx:    stx,
+		cache:  cache,
+		config: config,
+	}
+
+	return treeTX, nil
+}
+
+// getLatestRoot populates this TX with the newest tree root visible by this
+// transaction.
+func (t *treeTX) getLatestRoot(ctx context.Context) error {
+	t.getLatestRootOnce.Do(func() {
+		t._currentSTH, t._currentSTHErr = t.ts.latestSTH(ctx, t.stx, t.treeID)
+		if t._currentSTH != nil {
+			t._writeRev = t._currentSTH.TreeRevision + 1
+		}
+	})
+
+	return t._currentSTHErr
+}
+
+// treeTX is a concrete implementation of the Trillian
+// storage.TreeTX interface.
+type treeTX struct {
+	treeID int64
+
+	ts *treeStorage
+
+	// mu guards the nil setting/checking of stx as part of the open checking.
+	mu sync.RWMutex
+	// stx is the underlying Spanner transaction in which all operations will be
+	// performed.
+	stx spanRead
+
+	// config holds the StorageSettings proto acquired from the trillian.Tree.
+	// Varies according to tree_type (LogStorageConfig vs MapStorageConfig).
+	config proto.Message
+
+	// currentSTH holds a copy of the latest known STH at the time the
+	// transaction was started, or nil if there was no STH.
+	_currentSTH    *spannerpb.TreeHead
+	_currentSTHErr error
+
+	// writeRev is the tree revision at which any writes will be made.
+	_writeRev int64
+
+	cache cache.SubtreeCache
+
+	getLatestRootOnce sync.Once
+}
+
+func (t *treeTX) currentSTH(ctx context.Context) (*spannerpb.TreeHead, error) {
+	if err := t.getLatestRoot(ctx); err != nil {
+		return nil, err
+	}
+	return t._currentSTH, nil
+}
+
+func (t *treeTX) writeRev(ctx context.Context) (int64, error) {
+	if err := t.getLatestRoot(ctx); err != nil {
+		return -1, err
+	}
+	return t._writeRev, nil
+}
+
+// storeSubtrees adds buffered writes to the in-flight transaction to store the
+// passed in subtrees.
+func (t *treeTX) storeSubtrees(sts []*storagepb.SubtreeProto) error {
+	stx, ok := t.stx.(*spanner.ReadWriteTransaction)
+	if !ok {
+		return ErrWrongTXType
+	}
+	for _, st := range sts {
+		if st == nil {
```

Comment:
Can this legitimately be nil?

Reply:
Unsure, but I think similar checks are done elsewhere... Might be worth digging into at some point.